feat: (re)implement ASHA and set as default adaptive searcher [DET-3125]

liamcli opened the pull request:
## Description (WIP)
This PR implements an improved version of asynchronous successive halving.  The search algorithm promotes trials to higher rungs as they complete instead of waiting for all trials in a rung to complete before making promotions.  Adaptive now calls the new ASHA implementation with maxTrials but gives the user more control over the behavior while adaptive_simple uses default values for the divisor and brackets.
## Outstanding tasks
- [x] Add arg/config setting to control number of initialOperations of ASHA (which also controls the maximum parallelism of the experiment)
- [ ] Update documentation.
- [ ] Update examples.

## Test Plan
- [x] Update tests for ASHA.  
- [x] Update tests for adaptive and adaptive_simple.
- [x] Update test for tournament.  This test checks the asynchronous behavior of ASHA.
- [x] Test resume functionality of ASHA.

## Commentary (optional)
I could use help and advice for how best to add an arg that is a function of experiment max_slots and slots_per_trial to control degree of parallelism for ASHA (see first outstanding task).  Does it make sense to add it as a field to top level searcher config that is set in experiment.go and is ignored by most searchers by used by ASHA?

<!---
## Title

Example title: "docs: tweak recommended "pip install" usage".

Specifically, this title should contain a type and a description
of the change being made:

User-facing change types:
- docs: docs-only change
- feat: new user-facing feature
- fix: bug fix
- perf: performance improvement

Internal change types:
- build: build system change (anything in a `Makefile`, mostly)
- chore: any internal change not covered by another type
- ci: anything that touches `.circleci`
- refactor: internal refactor
- style: style change
- test: new tests

See https://www.conventionalcommits.org/en/v1.0.0/ for background.

The first line should also:
- be at most 89 characters long
- contain a description that is at most 72 characters long
- not end with sentence-ending punctuation
- start (after the type) with a lowercase imperative ("add", "fix")

## Description

Lead with the intended commit body in this description field. For breaking changes, please include "BREAKING CHANGE:" at the beginning of your commit body. At minimum, this section should include a bracketed reference to the Jira ticket, e.g. "[DET-1234]". When squash-and-merging, copy this directly into the description field.

## Test Plan

Describe the situations in which you've tested your change, and/or a screenshot as appropriate. Reviewers may ask questions of this test plan to ensure adequate manual coverage of changes.

## Commentary (optional)

Use this section of your description to add context to the PR. Could be for particularly tricky bits of code that could use extra scrutiny, historical context useful for reviewers, etc. You may intentionally leave this section blank and remove the title.
--->

[DET-1234]: https://determinedai.atlassian.net/browse/DET-1234

liamcli commented:
The go-lint check errored out for test code that it thought was duplicated but the code is actually different.  How can I remove this error?

dzhu commented:
You can add ` //nolint:dupl` at the beginning of each range that it thinks is duplicated.

brain-good commented:
@liamcli should we add the `WorkInProgress` label to the PR since it sounds like there are some open questions?

liamcli commented:
> @liamcli should we add the `Work in progress` label to the PR since it sounds like there are some open questions?
I changed the status of the PR to draft and added WIP to the description.  Let me know if there's another tag/flag you had in mind.

brain-good commented on master/pkg/searcher/asha.go:
```diff
@@ -1,169 +1,181 @@
-		return []RequestID{t.requestID}
 	}
+	if numPromote != oldNumPromote {
+		t := &r.metrics[oldNumPromote]
```
question: I'm really not sure about this logic, but granted my understanding of this algorithm is limited. It seems like we should be promoting the right most element of the array that has not already been promoted.

liamcli replied:
yes, we are doing that since things are zero indexed.  before trials from 0 to oldNumPromote-1 have been promoted so now the trial at index oldNumPromote needs to be promoted if possible.

brain-good commented on master/pkg/searcher/asha.go:
```diff
@@ -1,169 +1,181 @@
+	stepsNeeded       int
+	metrics           []trialMetric
+	startTrials       int
+	outstandingTrials int
```
question: do we need this? Shouldn't `outstandingTrials == startTrials - len(metrics)`?

liamcli replied:
startTrials is not a counter and is an int used for initialOperations.  len(metrics) is completed trials in a rung but not pending so that's why outstandingTrials is needed.

brain-good commented on master/pkg/searcher/asha.go:
```diff
@@ -1,169 +1,181 @@
 		ops = append(ops, create)
 		ops = append(ops, trainAndValidate(create.RequestID, 0, s.rungs[0].stepsNeeded)...)
+		s.rungs[0].outstandingTrials++
+		s.trialRungs[create.RequestID] = 0
```
question: why is this now needed?

liamcli replied:
trialRungs for a request needs to be initialized at the same time as trial creation since it is needed as part of validationCompleted.  oustandingTrials is used to track pending jobs so we know when all the jobs in a rung are complete and can close them and ultimately finish the experiment.

liamcli commented on master/pkg/searcher/asha.go:
```diff
@@ -1,169 +1,181 @@
-	metric, err := metrics.Metric(s.Metric)
+	metric, err := metrics.Metric(s.AsyncHalvingConfig.Metric)
+	// TODO: For graceful trial termination, if there is an error here, should we still
+	// continue the searcher and always return a new trial?
```
The searcher stops creating new trials once the trialRungs length matches maxTrials so it doesn't continue indefinitely.  In order to keep the worker parallelism the same for the experiment thought, we need to start a new job when a job errors out.  The question is, if an error is thrown here, do we then add an early exit event to the searcher operations?

liamcli commented on master/pkg/searcher/adaptive.go:
```diff
@@ -15,19 +41,22 @@ func newAdaptiveSearch(config model.AdaptiveConfig) SearchMethod {
-			TrainStragglers:  config.TrainStragglers,
 		}
 		methods = append(methods, newAsyncHalvingSearch(c))
+		fmt.Printf("Bracket created with %d rungs and %d max trials\n", numRungs, bracketMaxTrials[i])
```
which logging object should I pass to?  i see that there is a log in the actor context, is there one for master?  should we pass the log to the searcher context?

liamcli commented on master/pkg/searcher/adaptive_simple.go:
```diff
@@ -1,73 +1,32 @@
 func newAdaptiveSimpleSearch(config model.AdaptiveSimpleConfig) SearchMethod {
 	brackets := parseAdaptiveMode(config.Mode)(config.MaxRungs)
 	sort.Sort(sort.Reverse(sort.IntSlice(brackets)))
+	bracketMaxTrials := getBracketMaxTrials(config.MaxTrials, 4, brackets)
```
done.

liamcli commented on master/pkg/searcher/asha.go:
```diff
@@ -173,46 +185,71 @@ func (s *asyncHalvingSearch) promoteTrials(
+				// We make a recursive call that will behave the same
+				// as if we'd actually run the promoted job and received
+				// the worse possible result in return.
+				return s.promoteAsync(ctx, promotionID, wkld, ashaExitedMetricValue)
```
promoteAsync should not error, I've updated the definition to reflect that and changed the returns for validationCompleted and trialExitedEarly accordingly.

liamcli commented on master/pkg/searcher/asha.go:
```diff
@@ -173,46 +185,71 @@ func (s *asyncHalvingSearch) promoteTrials(
 				if !s.earlyExitTrials[trialMetric.requestID] {
 					ops = append(ops, NewClose(trialMetric.requestID))
 				}
+				// TODO: is checkpoint storage associated with the close operation?
```
deleting this since you answered them during our meeting on monday

liamcli commented on master/pkg/searcher/asha.go:
```diff
@@ -61,9 +58,9 @@ type rung struct {
-	startTrials       int
 	outstandingTrials int
 	promoteTrials     int
+	startTrials       int
```
startTrials is still used by synchronous SHA and the two share rung type.

brain-good commented:
@liamcli can you update the branch this is going to be merged into? This PR is not easily reviewable with all these changes in it

liamcli commented:
need to reopoen
